File: backend/src/main/java/com/ll/TeamProject/domain/calendar/entity/Calendar.java
package com.ll.TeamProject.domain.calendar.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.ll.TeamProject.domain.calendar.dto.CalendarUpdateDto;
import com.ll.TeamProject.domain.schedule.entity.Schedule;
import com.ll.TeamProject.domain.user.entity.SiteUser;
import com.ll.TeamProject.global.jpa.entity.BaseTime;
import jakarta.persistence.*;
import lombok.*;

import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
public class Calendar extends BaseTime {

    // 캘린더 소유 사용자
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private SiteUser user;

    // 캘린더 이름
    private String name;

    // 캘린더 설명
    private String description;

    // 공유된 사용자 목록
    @OneToMany(mappedBy = "calendar", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnore
    private List<SharedCalendar> sharedUsers = new ArrayList<>();

    // 메시지 목록
    @OneToMany(mappedBy = "calendar", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnore
    private List<Message> messageList = new ArrayList<>();

    // 일정 목록
    @OneToMany(mappedBy = "calendar", cascade = CascadeType.REMOVE)
    @JsonIgnore
    private List<Schedule> schedules = new ArrayList<>();

    // 비즈니스 생성자
    public Calendar(SiteUser user, String name, String description) {
        this.user = user;
        this.name = name;
        this.description = description;
    }

    // 캘린더 정보 업데이트 (record 적용)
    public void update(CalendarUpdateDto updateDto) {
        this.name = updateDto.name();
        this.description = updateDto.description();
    }

    // 캘린더 이름 및 설명 변경
    public void update(String name, String description) {
        this.name = name;
        this.description = description;
    }
}

File: backend/src/main/kotlin/com/ll/TeamProject/domain/calendar/entity/Calendar.kt
package com.ll.TeamProject.domain.calendar.entity

import com.fasterxml.jackson.annotation.JsonIgnore
import com.ll.TeamProject.domain.calendar.dto.CalendarUpdateDto
import com.ll.TeamProject.domain.schedule.entity.Schedule
import com.ll.TeamProject.domain.user.entity.SiteUser
import com.ll.TeamProject.global.jpa.entity.BaseTime
import jakarta.persistence.*

@Entity
class Calendar(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    var user: SiteUser, // 캘린더 소유자

    var name: String, // 캘린더 이름

    var description: String // 캘린더 설명
) : BaseTime() {

    // ✅ 공유된 사용자 목록 (캘린더 소유자 포함)
    @OneToMany(mappedBy = "calendar", cascade = [CascadeType.ALL], orphanRemoval = true)
    @JsonIgnore
    val sharedUsers: MutableList<SharedCalendar> = mutableListOf()

    // 일정 목록
    @OneToMany(mappedBy = "calendar", cascade = [CascadeType.REMOVE])
    @JsonIgnore
    val schedules: MutableList<Schedule> = mutableListOf()

    // ✅ 캘린더 정보 업데이트
    // 캘린더 정보 업데이트
    fun update(updateDto: CalendarUpdateDto) {
        this.name = updateDto.name
        this.description = updateDto.description
    }

    // ✅ 특정 사용자에게 캘린더 공유 (소유자 정보 포함)
    fun addSharedUser(user: SiteUser, owner: SiteUser) {
        if (sharedUsers.none { it.user == user }) {
            sharedUsers.add(SharedCalendar(this, user, owner)) // ✅ owner 명시적으로 전달
        }
    }

    // ✅ 특정 사용자와의 공유 해제
    fun removeSharedUser(user: SiteUser) {
        sharedUsers.removeIf { it.user == user }
    }

    // 기본 생성자
    constructor() : this(SiteUser(), "Default Name", "")
}
